### PR TITLE
Add additional open container api to HumanEntity

### PR DESCRIPTION
**Pull request by JRoy**
another round of "i want this in essentials but i don't want to enter nms hell"

### CODE REVIEW

File: Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+@@ -450,6 +450,116 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
+         return this.getHandle().activeContainer.getBukkitView();
+     }
+ 
++    // Paper start - Add additional containers
++    @Override
++    public InventoryView openAnvil(Location location, boolean force) {
++        if (!force) {
++            Block block = location.getBlock();
```
Comment (kashike) [inline Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch]: You check if`location` is `null` below, but not here - is this intentional? If so, why?

File: Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
++        if (force) {
++            getHandle().activeContainer.checkReachable = false;
++        }
++        return getHandle().activeContainer.getBukkitView();
++    }
++
++    @Override
++    public InventoryView openCartographyTable(Location location, boolean force) {
++        if (!force) {
++            Block block = location.getBlock();
```
Comment (kashike) [inline Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch]: You check if`location` is `null` below, but not here - is this intentional? If so, why?

File: Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
++        if (force) {
++            getHandle().activeContainer.checkReachable = false;
++        }
++        return getHandle().activeContainer.getBukkitView();
++    }
++
++    @Override
++    public InventoryView openGrindstone(Location location, boolean force) {
++        if (!force) {
++            Block block = location.getBlock();
```
Comment (kashike) [inline Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch]: You check if`location` is `null` below, but not here - is this intentional? If so, why?

File: Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
++            getHandle().activeContainer.checkReachable = false;
++        }
++        return getHandle().activeContainer.getBukkitView();
++    }
++
++    @Override
++    public InventoryView openLoom(Location location, boolean force) {
++        if (!force) {
++            Block block = location.getBlock();
++            if (block.getType() != Material.LOOM) {
```
Comment (kashike) [inline Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch]: You check if`location` is `null` below, but not here - is this intentional? If so, why?

File: Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
++        if (force) {
++            getHandle().activeContainer.checkReachable = false;
++        }
++        return getHandle().activeContainer.getBukkitView();
++    }
++
++    @Override
++    public InventoryView openSmithingTable(Location location, boolean force) {
++        if (!force) {
++            Block block = location.getBlock();
```
Comment (kashike) [inline Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch]: You check if`location` is `null` below, but not here - is this intentional? If so, why?

File: Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
++        if (force) {
++            getHandle().activeContainer.checkReachable = false;
++        }
++        return getHandle().activeContainer.getBukkitView();
++    }
++
++    @Override
++    public InventoryView openStonecutter(Location location, boolean force) {
++        if (!force) {
++            Block block = location.getBlock();
```
Comment (kashike) [inline Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch]: You check if`location` is `null` below, but not here - is this intentional? If so, why?

**Review by kashike — COMMENTED** [review]
It feels like a lot of the implementation duplicates code - you should be able to reduce it.

File: Spigot-API-Patches/0221-Add-additional-open-container-api-to-HumanEntity.patch [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
++     *
++     * @param location The location to attach it to. If null, the player's
++     *     location is used.
++     * @param force If false, and there is no cartography table block at the location,
++     *     no inventory will be opened and null will be returned.
++     * @return The newly opened inventory view, or null if it could not be
++     *     opened.
++     */
++    @Nullable
++    public InventoryView openCartographyTable(@Nullable Location location, boolean force);
```
Comment (Proximyst) [inline Spigot-API-Patches/0221-Add-additional-open-container-api-to-HumanEntity.patch]: Wouldn't it be easier for users to just specify inventory type?

File: Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
++            if (block.getType() != Material.STONECUTTER) {
++                return null;
++            }
++        }
++        if (location == null) {
++            location = getLocation();
++        }
++        getHandle().openContainer(((net.minecraft.server.BlockStonecutter) Blocks.STONECUTTER).getInventory(null, getHandle().world, new BlockPosition(location.getBlockX(), location.getBlockY(), location.getBlockZ())));
++        if (force) {
++            getHandle().activeContainer.checkReachable = false;
```
Comment (Proximyst) [inline Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch]: Just set to `!force`

File: Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
++        }
++        getHandle().openContainer(((net.minecraft.server.BlockSmithingTable) Blocks.SMITHING_TABLE).getInventory(null, getHandle().world, new BlockPosition(location.getBlockX(), location.getBlockY(), location.getBlockZ())));
++        if (force) {
++            getHandle().activeContainer.checkReachable = false;
++        }
++        return getHandle().activeContainer.getBukkitView();
++    }
++
++    @Override
++    public InventoryView openStonecutter(Location location, boolean force) {
```
Comment (Proximyst) [inline Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch]: Async catcher; this modifies state.

**Review by Proximyst — CHANGES_REQUESTED** [review]

File: Spigot-API-Patches/0221-Add-additional-open-container-api-to-HumanEntity.patch (hunk 7 above)
Comment (JRoy) [reply Spigot-API-Patches/0221-Add-additional-open-container-api-to-HumanEntity.patch]: Existing API to do this doesn't work correctly for these use cases (UnsupportedOperationExceptions + improper reachability checks). Due to the way containers are written, speific casting is required in order to properly so one general purpose method would end up being as long as indivial methods.

File: Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch (hunk 1 above)
Comment (JRoy) [reply Spigot-Server-Patches/0565-Add-additional-open-container-api-to-HumanEntity.patch]: I used code from the existing `openEnchant` method as a reference. I didn't notice this and it appears that this is a regression in the craftbukkit implementation.

Will fix this in these new methods

**Comment by JRoy** [comment]
@kashike Code duplication has been reduced as much as possible.

Due to the way these blocks are written, there is a lot of manual casting needed so that is why I had to shove a switch case in the helper method. Let me know if you want me to do that in a different way.

**Review by Proximyst — DISMISSED** [review]
I'll merge tomorrow if no-one else beats me to it.

**Review by aikar — DISMISSED** [review]

**Comment by MiniDigger** [comment]
@limbo-app ready

**Review by limbo-app — APPROVED** [review]
Pull request automatically approved for merging - a core team member has previously approved this pull request.

**Comment by MiniDigger** [comment]
thanks!

**Pull request merged**